managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java

```diff
@@ -1141,6 +1143,48 @@ public long getEstimatedBacklogSize() {
         }
     }
 
+    @Override
+    public long getEarliestMessagePublishTimeInBacklog() {
+        PositionImpl pos = getMarkDeletePositionOfSlowestConsumer();
+
+        return getEarliestMessagePublishTimeOfPos(pos);
+    }
+
+    public long getEarliestMessagePublishTimeOfPos(PositionImpl pos) {
+        if (pos == null) {
+            return 0L;
+        }
+        PositionImpl nextPos = getNextValidPosition(pos);
+
+        CompletableFuture<Long> future = new CompletableFuture<>();
+        asyncReadEntry(nextPos, new ReadEntryCallback() {
+            @Override
+            public void readEntryComplete(Entry entry, Object ctx) {
```

Comment:
We'd better check whether `broker_timestamp` has been set first, the priority of the timestamp is `broker_timestamp` > `publish_time` > `0`

Reply:
Fixed

Reply:
Maybe you could use the method `getEntryTimestamp()` of the class `MessageImpl`.

Reply:
> Maybe you could use the method `getEntryTimestamp()` of the class `MessageImpl`.

Fixed